File: 2_Periodo/Algoritmos_e_Estruturas_de_Dados_2/TPs/TP01/TP01Q10/PalindromoRec.java
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q10 - 11 / 03 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : javac PalindromoRec.java
 *       Windows: javac PalindromoRec.java
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : java PalindromoRec
 *       Windows: java PalindromoRec
 *   
*/

class PalindromoRec
{
    /**
     *  Funcao para verificar se a entrada e' igual a "FIM" 
     *  @param s - String
     *  @return true se fim, false caso contrario.
     */
    public static boolean isFim( String s )
    {
        boolean result = false;
        if( s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M' )
        {
            result = true;
        } // end if
        return ( result );
    } // end isFim ( )

    /**
     *  Funcao para verificar se uma string e' um palindromo.
     *  @return true se palindromo, false caso contrario.
     *  @param s - String
     *  @param x - Int: primeiro caractere.
     *  @param y - Int: ultimo caractere.
    */
    public static boolean isPalindromoRec ( String s, int x, int y )
    {
        if( x < s.length()/2 )
        {
            if( (int)s.charAt(x) == (int)s.charAt(y) )
            {
                return ( isPalindromoRec(s, ++x, --y) );
            }
            else
            {
                return ( false );
            } // end if
        } // end if
        return ( true );
    } // end isPalindromo ( )

    /**
     *  Funcao para verificar se uma string e' um palindromo.
     *  @return true se palindromo, false caso contrario.
     *  @param s - String
    */
    public static boolean isPalindromo( String s )
    {
        return ( isPalindromoRec( s, 0, s.length()-1 ) );
    } // end isPalindromo ( )

    /**
     *  Funcao principal.
     *  @param args
     */
    public static void main( String[] args ) 
    {
        MyIO.setCharset("UTF-8");
        String entrada = "";

        do 
        {
            entrada = MyIO.readLine();
            if( !isFim( entrada ) )
            {
                if( isPalindromo( entrada ) )
                {
                    MyIO.println( "SIM" );
                }
                else
                {
                    MyIO.println( "NAO" );
                } // end if
            } // end if
        } while ( isFim( entrada ) == false ); // end do while
    } // end main ( )
} // end class

File: 2_Periodo/Algoritmos_e_Estruturas_de_Dados_2/TPs/TP01/TP01Q11/PalindromoRec.c
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q11 - 12 / 03 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : gcc -o PalindromoRec PalindromoRec.c
 *       Windows: gcc -o PalindromoRec PalindromoRec.c
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : ./PalindromoRec
 *       Windows: PalindromoRec
 *   
*/

// ---------------------------------------- Dependencias

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// ---------------------------------------- Metodos

/**
 *  Funcao para verificar se a entrada e' igual a "FIM" 
 *  @param s - char*.
 *  @return true se fim, false caso contrario.
 */
bool isFim( char* s )
{
    bool result = false;
    if( strcmp( s, "FIM" ) == 0 )
    {
        result = true;
    }
    return ( result );
} // end isFim ( )

/**
 *  Funcao para verificar se uma string e' um palindromo.
 *  @return true se palindromo, false caso contrario.
 *  @param s - char*: string.
 *  @param s - int: primeiro caractere.
 *  @param y - int: ultimo caractere.
*/
bool isPalindromoRec ( char *s, int x, int y )
{
    if( x < strlen(s)/2 )
    {
        if( s[x] == s[y] )
        {
            return ( isPalindromoRec(s, ++x, --y) );
        }
        else
        {
            return ( false );
        } // end if
    } // end if
    return ( true );
} // end isPalindromoRec ( )

/**
 *  Funcao para verificar se uma string e' um palindromo.
 *  @return true se palindromo, false caso contrario.
 *  @param s - char*: string.
*/
bool isPalindromo( char* s )
{
    return ( isPalindromoRec( s, 0, strlen(s)-1 ) );
} // end isPalindromo ( )

/**
 *  Funcao para ler linha da entrada padrao.
 *  @return string.
*/
char* readLine( )
{
    char *string = (char*) calloc ( 1000+1, sizeof(char) );
    if( string != NULL )
    {
        fgets( string, 1000+1, stdin );
        unsigned int length = strlen( string );
        // if( length > 0 && ( string[length-1] == '\n' || string[length-1] == '\r') )
        if( length > 0 && string[length-1] == '\n' )
        {
            string[length-1] = '\0';
        } // end if
    } // end if
    return ( string ); 
} // end readLine ( )

/**
 *  Funcao Principal
 *  @param agrc
 *  @param agrv
*/
int main( int argc, char const *argv[] )
{
    char* entrada = NULL;

    do
    {
        entrada = readLine( );

        if( !isFim( entrada ) )
        {
            if( isPalindromo( entrada ) )
            {
                printf( "SIM\n" );
            }
            else
            {
                printf( "NAO\n" );
            } // end if
        } // end if
        free( entrada );
    } while ( !isFim( entrada ) );
    
    return ( 0 );
} // end main ( )